README.md
CogniCity
===========
**Open Source GeoSocial Intelligence Framework**

####cognicity-reports-powertrack: NodeJS app to collect unconfirmed reports from Twitter via Gnip PowerTrack and send verification requests.

[![Build Status](https://travis-ci.org/smart-facility/cognicity-reports-powertrack.svg)](https://travis-ci.org/smart-facility/cognicity-reports-powertrack)

### About
Cognicity-reports-powertrack is the NodeJS reports module for the CogniCity framework, responsible for collecting relevant tweets via Gnip PowerTrack, and sending users verification messages via Twitter. For detailed framework documentation see [http://cognicity.info](http://cognicity.info).

### API Documentation
[http://cognicity.info/cognicity/api-docs/cognicity-reports-powertrack/index.html](http://cognicity.info/cognicity/api-docs/cognicity-reports-powertrack/index.html)

### Dependencies
* [NodeJS](http://nodejs.org) version 0.10.16 or compatible
* [PostgreSQL](http://www.postgresql.org) version 9.2 or later, with [PostGIS](http://postgis/) version 2.0 or compatible

#### Node Modules
* [Node-Postgres](https://github.com/brianc/node-postgres) version 2.0.0 or compatible
* [ntwitter](https://github.com/AvianFlu/ntwitter) version 0.5.0 or compatible
* [gnip](https://github.com/demian85/gnip) version 0.2.1 or compatible
* [winston](https://github.com/flatiron/winston) version 0.8.1 or compatible

#### Dev Modules
* [jshint](https://github.com/jshint/node-jshint) version 2.5.8 or compatible
* [unit.js](http://unitjs.com/) version 1.0.2 or compatible
* [mocha](http://mochajs.org/) version 2.0.1 or compatible
* [jsdoc](https://github.com/jsdoc3/jsdoc) version 3.2.0 or compatible
* [istanbul](https://github.com/gotwarlost/istanbul) version 0.3.5 or compatible

If you're going to commit changes to the JavaScript, be sure to run 'npm test' first - and fix any issues that it complains about, otherwise the build will fail when you push the commit.

### Installation
Download the source code for cognicity-reports-powertrack from github: [http://github.com/smart-facility/cognicity-reports-powertrack](http://github.com/smart-facility/cognicity-reports-powertrack) or view the CogniCity installation documentation at [http://cognicity.info](http://cognicity.info).

Install the node dependencies in package.json using NPM: `npm install`

#### Platform-specific notes ####
To build on OS X we recommend using [homebrew](http://brew.sh) to install node, npm, and required node modules as follows:
```shell
brew install node
npm install
```

To build on Windows we recommend installing all dependencies (making sure to use all 32 bit or all 64 bit, depending on your architecture) plus following the instructions (for Windows 7 follow the Windows 7/8 instructions) for [node-gyp](https://github.com/TooTallNate/node-gyp) and then:
* You need to add *C:\Program Files\PostgreSQL\9.3\bin* (modifying that location if necessary to point to the installed version of PostgreSQL) to path so the build script finds `pg_config`, and
* You need to create the *%APPDATA%\npm* folder and run cmd (and hence npm) as administrator. *%APPDATA%* is usually under *C:\Users\your_username\AppData\Remote*.
* You may need to specify the version of the build tools installed by adding the argument `--msvs_version=2013` to the `npm` command (substituting the appropriate version)
Then you can run `npm install`.

### Configuration
App configuration parameters are stored in a configuration file which is parsed by app.js. See sample-reports-config.js for an example configuration.

#### Logging parameters
* level - info or debug are most useful here, debug will give you more verbose logging output
* maxFileSize - max size (in bytes) of each log file before a new one is created
* maxFiles - number of log files to retain
* logDirectory - Specify a full path to the log directory. If not specified, the application directory will be used.

#### Gnip parameters
* stream [true | false] - set to true to connect to Gnip stream.
* streamTimeout - Gnip stream timeout, should be >30s (in milliseconds)
* username - Gnip username
* password - Gnip password
* streamUrl - URL to fetch JSON stream from Gnip PowerTrack
* rulesUrl - URL to fetch JSON rules from Gnip PowerTrack
* rules - List of objects to configure Gnip PowerTrack rules. Objects contain a series of key-value properties, where the key is the Gnip PowerTrack tag for the rule, and the value is the rule as a string.
* maxReconnectTimeout - Time in seconds that is the longest delay between reconnection attempts for the stream

#### Twitter account configuration
Set the app authentication parameters as provided by Twitter. See the [ntwitter-module](https://github.com/AvianFlu/ntwitter) documentation for more details.
* usernameReplyBlacklist - Twitter usernames (without @, comma separated for multiples) which will never be sent to in response to tweet processing

#### Twitter warning configuration
* adminTwitterUsernames - Enter twitter usernames here (without @, comma separated for multiples) to send a notification tweet on error conditions
Notification conditions are:
* Gnip stream is disconnected and reaches `maxReconnectTimeout` time trying to reconnect
* Connection to postgres is lost and cannot be reconnected in `pg.reconnectionAttempts` number of attempts

#### Twitter send parameters
* send_enabled [true | false] - set to true to enable confirmation request tweets to be sent.
* addTimestamp [true | false] - if true, append a timestamp to each sent tweet.

#### Twitter message text
The messages are stored in objects, where the object name is the name of the message.
Within the object, the property name (key) is the language, and the value is the message text.
There is also a top-level 'defaultLanguage' property which is used if the language code from the tweet cannot be resolved.

##### Languages
* in - Bahasa Indonesian (language code from Gnip)
* id - Bahasa Indonesian (language code from Twitter)
* en - English

##### Messages
Messages can be at most 109 characters long if addTimestamp is enabled, or 123 characters long if addTimestamp is disabled.
* invite_text - Text for confirmation request tweets
* askforgeo_text - Text for geolocation reminders
* thanks_text - Thank-you message for confirmed tweet

#### Postgres connection
* conString - PostgreSQL connection details string (see node-postgres module documenation)[https://github.com/brianc/node-postgres]
* postgres tables as defined in database schema
* reconnectionDelay - Delay between reconnection attempts if postgres connection lost
* reconnectionAttempts - Number of times to attempt to reconnect before dying

### PostgreSQL/PostGIS schema
* see the [cognicity-schema](https://github.com/smart-facility/cognicity-schema) project for schema files

### Run
The app can be run as a background process using the [pm2 process manager](https://github.com/Unitech/pm2).

To install pm2, run:
```shell
sudo npm install pm2 -g
```
The app can then be started using
```shell
pm2 start processes.json
```
To have pm2 started on OS startup run
```shell
pm2 startup
```
and then run the command as per the instructions that prints out. If that command errors then you may have to specify the system (note that systemd should be used on CentOS 7). Note that if the process is not running as root (recommended) you will need to change `/etc/init.d/pm2-init.sh` to set `export PM2_HOME="/path/to/user/home/.pm2"`, as per [these instructions](
http://www.buildsucceeded.com/2015/solved-pm2-startup-at-boot-time-centos-7-red-hat-linux/)

The file [processes.json](processes.json) contains a number of options that can be set, including the name of the process (default: "harvester") and the watch list. At the moment any paths or files starting with . (including .git), node_modules, git, test folders, and all \*.log files will be ignored, but any other changes (e.g. to a config file, or to the code itself) will automatically result in a restart of the process. Refer to the [documentation](http://pm2.keymetrics.io/docs/usage/application-declaration/) for more options in the [processes.json](processes.json) file.

For further details refer to the [README for pm2](https://github.com/Unitech/PM2/blob/master/README.md).

### Logging
* Winston writes to project-name.log (and project-name#.log if configured for multiple files)
* The log directory can be configured, by default it is the project directory

### Development

#### Testing

To run the full set of tests, run:

```shell
npm test
```

This will run the following tests:

##### JSHint

JSHint will run on all JavaScript files in the root folder and test folders.

Running the script:

```shell
npm run jshint
```

This will print an error to the screen if any problems are found.

##### Mocha

Mocha will run all unit tests in the test folder and can be run with the following script:

```shell
npm run mocha
```

The test output will tell you how many tests passed and failed.

#### Git Hooks

There is a git pre-commit hook which will run the 'npm test' command before your commit and will fail the commit if testing fails.

To use this hook, copy the file from 'git-hooks/pre-commit' to '.git/hooks/pre-commit' in your project folder.

```shell
cp git-hooks/pre-commit .git/hooks/
```

#### Documentation

To build the JSDoc documentation run the following npm script:

```shell
npm run build-docs
```

This will generate the API documentation in HTML format in the `docs` folder, where you can open it with a web browser.

#### Test Coverage

To build test code coverage documentation, run the following npm script:

```shell
npm run coverage
```

This will run istanbul code coverage over the full mocha test harness and produce HTML documentation in the directory `coverage` where you can open it with a web browser.

#### Release

The release procedure is as follows:
* Update the CHANGELOG.md file with the newly released version, date, and a high-level overview of changes. Commit the change.
* Create a tag in git from the current head of master. The tag version should be the same as the version specified in the package.json file - this is the release version.
* Update the version in the package.json file and commit the change.
* Further development is now on the updated version number until the release process begins again.

### License
This software is released under the GPLv3 License. See License.txt for details.


CHANGELOG.md
CogniCity
===========
**Open Source GeoSocial Intelligence Framework**

####cognicity-reports-powertrack Changelog

### 1.0.0 - 24/11/2014
* First release

### 1.0.1 - 17/12/2014
* Handle temporary Postgres downtime (tweets cached)
* Fix for always sending geo-tweets without recursion from whitelisted account
* Improved error checking at start-up
* Tweet admin account on un-handled exceptions

### 1.0.2 - 05/01/2015
* Blacklist of usernames not to reply tweet to available
* Log DB connection success message at startup

### 1.0.3 - 21/01/2015
* Updated JSDoc outputs
* Added Istanbul code coverage.

### 1.0.4 - 08/10/2015
* Removed node-daemonize2 support
* Improved config file for PM2 support


test/test-config.js
'use strict';

// test-config.js - include the sample config and change some properties for test

var config = require("../sample-reports-config.js");

// Change instance name so we log to test.log
config.logger.filename = 'test';

// Export our modified config
module.exports = config;


CognicityReportsPowertrack.js
'use strict';

// CognicityReportsPowertrack.js - cognicity-reports-powertrack modules

/**
 * A CognicityReportsPowertrack object:
 * - connects to a powertrack stream
 * - monitors tweets based on configuration
 * - sends messages to users via twitter
 * - stores data from tweets in database
 * @constructor
 * @param {config} config Configuration object
 * @param {object} twit Configured instance of twitter object from ntwitter module
 * @param {object} pg Configured instance of pg object from pg module
 * @param {object} logger Configured instance of logger object from Winston module
 * @param {object} Gnip Instance of Gnip object from Gnip module
 */
var CognicityReportsPowertrack = function(
	config,
	twit,
	pg,
	logger,
	Gnip
		){

	this.config = config;
	this.twit = twit;
	this.pg = pg;
	this.logger = logger;
	this.Gnip = Gnip;
};

CognicityReportsPowertrack.prototype = {
	/**
	 * Configuration object
	 * @type {config}
	 */
	config: null,
	/**
	 * Configured instance of twitter object from ntwitter module
	 * @type {object}
	 */
	twit: null,
	/**
	 * Configured instance of pg object from pg module
	 * @type {object}
	 */
	pg: null,
	/**
	 * Configured instance of logger object from Winston module
	 * @type {object}
	 */
	logger: null,
	/**
	 * Instance of Gnip object from Gnip module
	 * @type {object}
	 */
	Gnip: null,
	/**
	 * Flag signifying if we are currently able to process tweets immediately.
	 * Turned on if the database is temporarily offline so we can cache tweets for a short time.
	 * @type {boolean}
	 */
	_cacheMode: false,
	/**
	 * Store tweets if we cannot process them immediately, for later processing.
	 * @type {Array}
	 */
	_cachedTweets: [],

	/**
	 * Resolve message code from config.twitter using passed language codes.
	 * Will fall back to trying to resolve message using default language set in configuration.
	 * @param {string} code Message code to lookup in config.twitter
	 * @param {GnipTweetActivity} tweetActivity Tweet activity object; check twitter language code and Gnip language codes for a match
	 * @returns {?string} Message text, or null if not resolved.
	 */
	getMessage: function(code, tweetActivity) {
		var self = this;

		// Fetch the language codes from both twitter and Gnip data, if present
		var langs = [];
		if (tweetActivity.twitter_lang) langs.push(tweetActivity.twitter_lang);
		if (tweetActivity.gnip && tweetActivity.gnip.language && tweetActivity.gnip.language.value) langs.push(tweetActivity.gnip.language.value);

		// Find a matching code if we can
		if (self.config.twitter[code]) {
			for (var i=0; i<langs.length; i++) {
				var lang = langs[i];
				if (self.config.twitter[code][lang]) return self.config.twitter[code][lang];
			}
			// If we haven't found a code, try the default language
			if (self.config.twitter[code][self.config.twitter.defaultLanguage]) return self.config.twitter[code][self.config.twitter.defaultLanguage];
		}

		self.logger.warn( "getMessage: Code could not be resolved for '" + code + "' and langs '" + langs +"'" );
		return null;
	},

	/**
	 * DB query success callback
	 * @callback DbQuerySuccess
	 * @param {object} result The 'pg' module result object on a successful query
	 */

	/**
	 * Gnip PowerTrack Tweet Activity object.
	 * @see {@link http://support.gnip.com/sources/twitter/data_format.html}
	 * @typedef GnipTweetActivity
	 */
	
	/**
	 * Execute the SQL against the database connection. Run the success callback on success if supplied.
	 * @param {object} config The pg config object for a parameterized query, e.g. {text:"select * from foo where a=$1", values:['bar']}
	 * @param {string} config.text The SQL query to execute
	 * @param {Array} config.values Parameterized values to use for the SQL query
	 * @param {DbQuerySuccess} success Callback function to execute on success.
	 */
	dbQuery: function(config, success){
		var self = this;

		self.logger.debug( "dbQuery: executing query: " + JSON.stringify(config) );
		self.pg.connect(self.config.pg.conString, function(err, client, done){
			if (err){
				self.logger.error("dbQuery: " + JSON.stringify(config) + ", " + err);
				done();
				return;
			}
			client.query(config, function(err, result){
				if (err){
					self.logger.error("dbQuery: " + JSON.stringify(config) + ", " + err);
					done();
					return;
				}
				done();
				self.logger.debug( "dbQuery: success: " + JSON.stringify(config) );
				if (success) {
					try {
						success(result);
					} catch(error) {
						self.logger.error("dbQuery: Error in success callback: " + error.message + ", " + error.stack);
					}
				}
			});
		});
	},

	/**
	 * Only execute the success callback if the user is not currently in the all users table.
	 * @param {string} user The twitter screen name to check if exists
	 * @param {DbQuerySuccess} callback Callback to execute if the user doesn't exist
	 */
	ifNewUser: function(user, success){
		var self = this;

		self.dbQuery(
			{
				text: "SELECT user_hash FROM " + self.config.pg.table_all_users + " WHERE user_hash = md5($1);",
				values: [ user ]
			},
			function(result) {
				if (result && result.rows && result.rows.length === 0) {
					success(result);
				} else {
					self.logger.debug("Not performing callback as user already exists");
				}
			}
		);
	},

	/**
	 * Send @reply Twitter message
	 * @param {GnipTweetActivity} tweetActivity The Gnip tweet activity object this is a reply to
	 * @param {string} message The tweet text to send
	 * @param {function} success Callback function called on success
	 */
	sendReplyTweet: function(tweetActivity, message, success){
		var self = this;
		
		var usernameInBlacklist = false;
		if (self.config.twitter.usernameReplyBlacklist) {
			self.config.twitter.usernameReplyBlacklist.split(",").forEach( function(blacklistUsername){
				if ( tweetActivity.actor.preferredUsername === blacklistUsername.trim() ) usernameInBlacklist = true;
			});
		}
		
		if ( usernameInBlacklist ) {
			// Never send tweets to usernames in blacklist
			self.logger.info( 'sendReplyTweet: Tweet user is in usernameReplyBlacklist, not sending' );
		} else {
			// Tweet is not to ourself, attempt to send
			var originalTweetId = tweetActivity.id;
			originalTweetId = originalTweetId.split(':');
			originalTweetId = originalTweetId[originalTweetId.length-1];

			var params = {};
			params.in_reply_to_status_id = originalTweetId;

			message = '@' + tweetActivity.actor.preferredUsername + ' ' + message;
			if ( self.config.twitter.addTimestamp ) message = message + " " + new Date().getTime();

			if (self.config.twitter.send_enabled === true){
				self.twit.updateStatus(message, params, function(err, data){
					if (err) {
						self.logger.error( 'Tweeting "' + message + '" with params "' + JSON.stringify(params) + '" failed: ' + err );
					} else {
						self.logger.debug( 'Sent tweet: "' + message + '" with params ' + JSON.stringify(params) );
						if (success) success();
					}
				});
			} else { // for testing
				self.logger.info( 'sendReplyTweet: In test mode - no message will be sent. Callback will still run.' );
				self.logger.info( 'sendReplyTweet: Would have tweeted: "' + message + '" with params ' + JSON.stringify(params) );
				if (success) success();
			}	
		}
	},

	/**
	 * Insert a confirmed report - i.e. has geo coordinates and is addressed.
	 * Store both the tweet information and the user hash.
	 * @param {GnipTweetActivity} tweetActivity Gnip PowerTrack tweet activity object
	 */
	insertConfirmed: function(tweetActivity){
		var self = this;

		//insertUser with count -> upsert
		self.dbQuery(
			{
				text : "INSERT INTO " + self.config.pg.table_tweets + " " +
					"(created_at, text, hashtags, urls, user_mentions, lang, the_geom) " +
					"VALUES (" +
					"$1, " +
					"$2, " +
					"$3, " +
					"$4, " +
					"$5, " +
					"$6, " +
					"ST_GeomFromText('POINT(' || $7 || ')',4326)" +
					");",
				values : [
				    tweetActivity.postedTime,
				    tweetActivity.body,
				    JSON.stringify(tweetActivity.twitter_entities.hashtags),
				    JSON.stringify(tweetActivity.twitter_entities.urls),
				    JSON.stringify(tweetActivity.twitter_entities.user_mentions),
				    tweetActivity.twitter_lang,
				    tweetActivity.geo.coordinates[1] + " " + tweetActivity.geo.coordinates[0]
				]
			},
			function(result) {
				self.logger.info('Logged confirmed tweet report');
				self.dbQuery(
					{
						text : "SELECT upsert_tweet_users(md5($1));",
						values : [
						    tweetActivity.actor.preferredUsername
						]
					},
					function(result) {
						self.logger.info('Logged confirmed tweet user');
						// Send the user a thank-you tweet; send this for every confirmed report
						self.sendReplyTweet( tweetActivity, self.getMessage('thanks_text', tweetActivity) );
					}
				);
			}
		);
	},

	/**
	 * Insert an invitee - i.e. a user we've invited to participate.
	 * @param {GnipTweetActivity} tweetActivity Gnip PowerTrack tweet activity object
	 */
	insertInvitee: function(tweetActivity){
		var self = this;

		self.dbQuery(
			{
				text : "INSERT INTO " + self.config.pg.table_invitees + " (user_hash) VALUES (md5($1));",
				values : [ tweetActivity.actor.preferredUsername ]
			},
			function(result) {
				self.logger.info('Logged new invitee');
			}
		);
	},

	/**
	 * Insert an unconfirmed report - i.e. has geo coordinates but is not addressed.
	 * @param {GnipTweetActivity} tweetActivity Gnip PowerTrack tweet activity object
	 */
	insertUnConfirmed: function(tweetActivity){
		var self = this;

		self.dbQuery(
			{
				text : "INSERT INTO " + self.config.pg.table_unconfirmed + " " +
					"(created_at, the_geom) " +
					"VALUES ( " +
					"$1, " +
					"ST_GeomFromText('POINT(' || $2 || ')',4326)" +
					");",
				values : [
				    tweetActivity.postedTime,
				    tweetActivity.geo.coordinates[1] + " " + tweetActivity.geo.coordinates[0]
				]
			},
			function(result) {
				self.logger.info('Logged unconfirmed tweet report');
			}
		);
	},

	/**
	 * Insert a non-spatial tweet report - i.e. we got an addressed tweet without geo coordinates.
	 * @param {GnipTweetActivity} tweetActivity Gnip PowerTrack tweet activity object
	 */
	insertNonSpatial: function(tweetActivity){
		var self = this;

		self.dbQuery(
			{
				text : "INSERT INTO " + self.config.pg.table_nonspatial_tweet_reports + " " +
					"(created_at, text, hashtags, urls, user_mentions, lang) " +
					"VALUES (" +
					"$1, " +
					"$2, " +
					"$3, " +
					"$4, " +
					"$5, " +
					"$6" +
					");",
				values : [
					tweetActivity.postedTime,
					tweetActivity.body,
					JSON.stringify(tweetActivity.twitter_entities.hashtags),
					JSON.stringify(tweetActivity.twitter_entities.urls),
					JSON.stringify(tweetActivity.twitter_entities.user_mentions),
					tweetActivity.twitter_lang
				]
			},

			function(result) {
				self.logger.info('Inserted non-spatial tweet');
			}
		);

		self.ifNewUser( tweetActivity.actor.preferredUsername, function(result) {
			self.dbQuery(
				{
					text : "INSERT INTO " + self.config.pg.table_nonspatial_users + " (user_hash) VALUES (md5($1));",
					values : [ tweetActivity.actor.preferredUsername ]
				},
				function(result) {
					self.logger.info("Inserted non-spatial user");
				}
			);
		});
	},

	/**
	 * Main stream tweet filtering logic.
	 * Filter the incoming tweet and decide what action needs to be taken:
	 * confirmed report, ask for geo, ask user to participate, or nothing
	 * @param {GnipTweetActivity} tweetActivity The tweet activity from Gnip
	 */
	filter: function(tweetActivity){
		var self = this;

		self.logger.verbose( 'filter: Received tweetActivity: screen_name="' + tweetActivity.actor.preferredUsername + '", text="' + tweetActivity.body.replace("\n", "") + '", coordinates="' + (tweetActivity.geo && tweetActivity.geo.coordinates ? tweetActivity.geo.coordinates[1]+", "+tweetActivity.geo.coordinates[0] : 'N/A') + '"' );

		// Everything incoming has a keyword already, so we now try and categorize it using the Gnip tags
		var hasGeo = (tweetActivity.geo && tweetActivity.geo.coordinates);
		var geoInBoundingBox = false;
		var addressed = false;
		var locationMatch = false;

		tweetActivity.gnip.matching_rules.forEach( function(rule){
			if (rule.tag) {
				// Only set geoInBoundingBox to true if the user has tweet coordinates as well as matching the rule,
				// as the rule can match on 'place' being within the bounding box
				if (rule.tag.indexOf("boundingbox")===0 && hasGeo) geoInBoundingBox = true;
				if (rule.tag.indexOf("addressed")===0) addressed = true;
				if (rule.tag.indexOf("location")===0) locationMatch = true;
			}
		});
		var tweetCategorizations = (geoInBoundingBox?'+':'-') + "BOUNDINGBOX " +
			(hasGeo?'+':'-') + "GEO " +
			(addressed?'+':'-') + "ADDRESSED " +
			(locationMatch?'+':'-') + "LOCATION";

		self.logger.verbose("filter: Categorized tweetActivity via Gnip tags as " + tweetCategorizations);

		// Perform the actions for the categorization of the tween
		if ( geoInBoundingBox && addressed ) {
			self.logger.verbose( 'filter: +BOUNDINGBOX +ADDRESSED = confirmed report' );

			self.insertConfirmed(tweetActivity); //user + geo = confirmed report!

		} else if ( geoInBoundingBox && !addressed ) {
			self.logger.verbose( 'filter: +BOUNDINGBOX -ADDRESSED = unconfirmed report, ask user to participate' );

			self.insertUnConfirmed(tweetActivity); //insert unconfirmed report

			// If we haven't contacted the user before, send them an invite tweet
			self.ifNewUser( tweetActivity.actor.preferredUsername, function(result) {
				self.sendReplyTweet(tweetActivity, self.getMessage('invite_text', tweetActivity), function(){
					self.insertInvitee(tweetActivity);
				});
			});

		} else if ( !geoInBoundingBox && !hasGeo && locationMatch && addressed ) {
			self.logger.verbose( 'filter: -BOUNDINGBOX -GEO +ADDRESSED +LOCATION = ask user for geo' );

			self.insertNonSpatial(tweetActivity); //User sent us a message but no geo, log as such

			// Ask them to enable geo-location
			self.sendReplyTweet( tweetActivity, self.getMessage('askforgeo_text', tweetActivity) );

		} else if ( !geoInBoundingBox && !hasGeo && locationMatch && !addressed ) {
			self.logger.verbose( 'filter: -BOUNDINGBOX -GEO -ADDRESSED +LOCATION = ask user to participate' );

			// If we haven't contacted the user beforem, send them an invite tweet
			self.ifNewUser( tweetActivity.actor.preferredUsername, function(result) {
				self.sendReplyTweet(tweetActivity, self.getMessage('invite_text', tweetActivity), function(){
					self.insertInvitee(tweetActivity);
				});
			});

		} else {
			// Not in bounding box but has geocoordinates or no location match
			self.logger.warn( 'filter: Tweet did not match category actions: ' + tweetCategorizations );
		}

	},

	/**
	 * Connect the Gnip stream.
	 * Establish the network connection, push rules to Gnip.
	 * Setup error handlers and timeout handler.
	 * Handle events from the stream on incoming data.
	 */
	connectStream: function(){
		var self = this;

		// Gnip stream
		var stream;
		// Timeout reconnection delay, used for exponential backoff
		var _initialStreamReconnectTimeout = 1000;
		var streamReconnectTimeout = _initialStreamReconnectTimeout;
		// Connect Gnip stream and setup event handlers
		var reconnectTimeoutHandle;
		// Send a notification on an extended disconnection
		var disconnectionNotificationSent = false;

		// Attempt to reconnect the socket.
		// If we fail, wait an increasing amount of time before we try again.
		function reconnectSocket() {
			// Try and destroy the existing socket, if it exists
			self.logger.warn( 'connectStream: Connection lost, destroying socket' );
			if ( stream._req ) stream._req.destroy();

			// If our timeout is above the max threshold, cap it and send a notification tweet
			if (streamReconnectTimeout >= self.config.gnip.maxReconnectTimeout) {
				// Only send the notification once per disconnection
				if (!disconnectionNotificationSent) {
					var message = "Cognicity Reports PowerTrack Gnip connection has been offline for " +
						self.config.gnip.maxReconnectTimeout + " seconds";
					self.tweetAdmin(message);
					disconnectionNotificationSent = true;
				}
			} else {
				streamReconnectTimeout *= 2;
				if (streamReconnectTimeout >= self.config.gnip.maxReconnectTimeout) streamReconnectTimeout = self.config.gnip.maxReconnectTimeout;
			}

			// Attempt to reconnect
			self.logger.info( 'connectStream: Attempting to reconnect stream' );
			stream.start();
		}

		// TODO We get called twice for disconnect, once from error once from end
		// Is this normal? Can we only use one event? Or is it possible to get only
		// one of those handlers called under some error situations.

		// Attempt to reconnect the Gnip stream.
		// This function handles us getting called multiple times from different error handlers.
		function reconnectStream() {
			if (reconnectTimeoutHandle) clearTimeout(reconnectTimeoutHandle);
			self.logger.info( 'connectStream: queing reconnect for ' + streamReconnectTimeout );
			reconnectTimeoutHandle = setTimeout( reconnectSocket, streamReconnectTimeout );
		}

		// Configure a Gnip stream with connection details
		stream = new self.Gnip.Stream({
		    url : self.config.gnip.streamUrl,
		    user : self.config.gnip.username,
		    password : self.config.gnip.password
		});

		// When stream is connected, setup the stream timeout handler
		stream.on('ready', function() {
			self.logger.info('connectStream: Stream ready!');
		    streamReconnectTimeout = _initialStreamReconnectTimeout;
		    disconnectionNotificationSent = false;
			// Augment Gnip.Stream._req (Socket) object with a timeout handler.
			// We are accessing a private member here so updates to gnip could break this,
		    // but gnip module does not expose the socket or methods to handle timeout.
			stream._req.setTimeout( self.config.gnip.streamTimeout, function() {
				self.logger.error('connectStream: Timeout error on Gnip stream');
				reconnectStream();
			});
		});

		// When we receive a tweetActivity from the Gnip stream this event handler will be called
		stream.on('tweet', function(tweetActivity) {
			if (self._cacheMode) {
				self.logger.debug( "connectStream: caching incoming tweet for later processing (id=" + tweetActivity.id + ")" );
				self._cachedTweets.push( tweetActivity );
			} else {
				self.logger.debug("connectStream: stream.on('tweet'): tweet = " + JSON.stringify(tweetActivity));

				// Catch errors here, otherwise error in filter method is caught as stream error
				try {
					if (tweetActivity.actor) {
						// This looks like a tweet in Gnip activity format
						self.filter(tweetActivity);
					} else {
						// This looks like a system message
						self.log.info("connectStream: Received system message: " + JSON.stringify(tweetActivity));
					}
				} catch (err) {
					self.logger.error("connectStream: stream.on('tweet'): Error on handler:" + err.message + ", " + err.stack);
				}
			}
		});

		// Handle an error from the stream
		stream.on('error', function(err) {
			self.logger.error("connectStream: Error connecting stream:" + err);
			reconnectStream();
		});

		// TODO Do we need to catch the 'end' event?
		// Handle a socket 'end' event from the stream
		stream.on('end', function() {
			self.logger.error("connectStream: Stream ended");
			reconnectStream();
		});

		// Construct a Gnip rules connection
		var rules = new self.Gnip.Rules({
		    url : self.config.gnip.rulesUrl,
		    user : self.config.gnip.username,
		    password : self.config.gnip.password
		});

		// Create rules programatically from config
		// Use key of rule entry as the tag, and value as the rule string
		var newRules = [];
		for (var tag in self.config.gnip.rules) {
			if ( self.config.gnip.rules.hasOwnProperty(tag) ) {
				newRules.push({
					tag: tag,
					value: self.config.gnip.rules[tag]
				});
			}
		}
		self.logger.debug('connectStream: Rules = ' + JSON.stringify(newRules));

		// Push the parsed rules to Gnip
		self.logger.info('connectStream: Updating rules...');
		// Bypass the cache, remove all the rules and send them all again
		rules.live.update(newRules, function(err) {
		    if (err) throw err;
			self.logger.info('connectStream: Connecting stream...');
			// If we pushed the rules successfully, now try and connect the stream
			stream.start();
		});

	},
	
	/**
	 * Stop realtime processing of tweets and start caching tweets until caching mode is disabled.
	 */
	enableCacheMode: function() {
		var self = this;
		
		self.logger.verbose( 'enableCacheMode: Enabling caching mode' );
		self._cacheMode = true;
	},
	
	/**
	 * Resume realtime processing of tweets.
	 * Also immediately process any tweets cached while caching mode was enabled.
	 */
	disableCacheMode: function() {
		var self = this;
		
		self.logger.verbose( 'disableCacheMode: Disabling caching mode' );
		self._cacheMode = false;
		
		self.logger.verbose( 'disableCacheMode: Processing ' + self._cachedTweets.length + ' cached tweets' );
		self._cachedTweets.forEach( function(tweetActivity) {
			self.filter(tweetActivity);
		});
		self.logger.verbose( 'disableCacheMode: Cached tweets processed' );
		self._cachedTweets = [];
	},
	
	/**
	 * Tweet the admin usernames defined in 'adminTwitterUsernames' in config.
	 * @param {string} warningMessage The message to tweet
	 * @param {function} callback Callback to execute after tweet sending
	 */
	tweetAdmin: function(warningMessage, callback) {
		var self = this;
		
		// Send notification tweet if we have a configured username
		if (self.config.adminTwitterUsernames) {
			// Construct the notification messages for each user.
			self.config.adminTwitterUsernames.split(",").forEach( function(username){
				var trimmedUsername = username.trim();
				// Always timestamp this, otherwise the messages will always look the same and won't post.
				var message = '@' + trimmedUsername +
					' ' + warningMessage +
					" " + new Date().getTime();

				self.logger.warn( 'tweetAdmin: Tweeting warning: "' + message + '"' );
				self.twit.updateStatus(message, function(err, data){
					if (err) self.logger.error('tweetAdmin: Tweeting failed: ' + err);
					if (callback) callback();
				});
			});
		}
	},
	
	/**
	 * Check that all tweetable message texts are of an acceptable length.
	 * This is 109 characters max if timestamps are enabled, or 123 characters max if timestamps are not enabled.
	 * @see {@link https://dev.twitter.com/overview/api/counting-characters} (max tweet = 140 chars)
	 * @see {@link https://support.twitter.com/articles/14609-changing-your-username} (max username = 15 chars)
	 * @return {boolean} True if message texts are all okay, false if any are not.
	 */
	areTweetMessageLengthsOk: function() {
		var self = this;
		var lengthsOk = true;
		
		Object.keys( self.config.twitter ).forEach( function(configItemKey) {
			// We only want to process the objects containing language/message pairs here,
			// not the single properties.
			var configItem = self.config.twitter[configItemKey];
			if (typeof configItem === "object") {
				var maxLength = 140; // Maximum tweet length
				maxLength -= 17; // Minus username, @ sign and space = 123
				if ( self.config.twitter.addTimestamp ) maxLength -= 14; // Minus 13 digit timestamp + space = 109 (13 digit timestamp is ok until the year 2286)
				Object.keys( configItem ).forEach( function(messageKey) {
					var message = configItem[messageKey];
					if ( message.length > maxLength ) {
						self.logger.error( "Message " + configItemKey + "." + messageKey + " '" + message + "' is too long (" + message.length + " chars)" );
						lengthsOk = false;
					}
				});
			}
		});
		
		return lengthsOk;
	}
};

// Export our object constructor method from the module
module.exports = CognicityReportsPowertrack;
